include/Application.h

```diff
@@ -21,8 +25,26 @@ namespace ccm
 
 		bool m_quit{ false };
 
+		template<typename T, typename... Args>
+		void replaceTopState(Args... args)
+		{
+			std::unique_ptr<GameState> nextState = std::make_unique<T>(args...);
+			popState()
+			m_gameStates.push_back(std::move(nextState));
+		}
+
+		template<typename T, typename... Args>
+		void pushState(Args... args)
+		{
+			std::unique_ptr<GameState> nextState = std::make_unique<T>(args...);
```

Comment:
I would use forwarding references which can be forwarded straight into the unique_ptr/vector. To perhaps cutdown on repetition, you could also do:
```cpp

template<typename T, typename... Args>
void pushState(Args&&... args)
{
    m_gameStates.push_back(std::make_unique<T>(std::forward<Args>(args)...));
}

template<typename T, typename... Args>
void replaceTopState(Args&&... args)
{
    popState();
    pushState<T>(std::forward<Args>(args)...);
}
```
since this is the job of pushState anyway.

Reply:
I'll change the references thing. Regarding the second bit, `replaceState()` needs to do its thing in that specific order, or if you call it from within the state that's getting replaced, it will pop itself (destroying the data) then possibly try to use that data in the arguments for the next state.

For example, say the GameplayState checked if the game was over inside it, and when the game was over, it replaces itself with a GameOverState. Let's also assume you need to tell the GameOverState whether you won or lost. If you pop first, the won/lost variable will be destroyed, then you try to pass it into the `unique_ptr`'s ctor. This was a pain to figure out lol